jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCookieStore.java

```diff
@@ -283,64 +304,97 @@ public List<HttpCookie> match(URI uri)
+            if (cookieDomain == null)
+                return uriDomain.equalsIgnoreCase(domain);
+            return isSameOrSubDomain(uriDomain, cookieDomain);
+        }
+
+        private static boolean isSameOrSubDomain(String subDomain, String domain)
+        {
+            int subDomainLength = subDomain.length();
+            int domainLength = domain.length();
+            // Case-insensitive version of subDomain.endsWith(domain).
+            if (!subDomain.regionMatches(true, subDomainLength - domainLength, domain, 0, domainLength))
+                return false;
+            // Make sure it is a subdomain.
+            int beforeMatch = subDomainLength - domainLength - 1;
+            // Domains are the same.
+            if (beforeMatch < 0)
                 return true;
-            if (domain != null)
-                domain = domain.toLowerCase(Locale.ENGLISH);
-            uriDomain = uriDomain.toLowerCase(Locale.ENGLISH);
-            if (cookieDomain != null)
-                cookieDomain = cookieDomain.toLowerCase(Locale.ENGLISH);
-            if (cookieDomain == null || cookieDomain.endsWith("."))
-            {
-                // RFC 6265, section 4.1.2.3.
-                // No cookie domain -> cookie sent only to origin server.
-                return uriDomain.equals(domain);
-            }
-            if (cookieDomain.startsWith("."))
-                cookieDomain = cookieDomain.substring(1);
-            if (uriDomain.endsWith(cookieDomain))
-            {
-                // The domain is the same as, or a subdomain of, the cookie domain.
-                int beforeMatch = uriDomain.length() - cookieDomain.length() - 1;
-                // Domains are the same.
-                if (beforeMatch == -1)
-                    return true;
-                // Verify it is a proper subdomain such as bar.foo.com,
-                // not just a suffix of a domain such as bazfoo.com.
-                return uriDomain.charAt(beforeMatch) == '.';
-            }
-            return false;
+            // Verify it is a proper subdomain such as bar.foo.com,
+            // not just a suffix of a domain such as bazfoo.com.
+            return subDomain.charAt(beforeMatch) == '.';
         }
 
-        private static boolean pathMatches(String path, String cookiePath)
+        private static boolean pathMatches(String uriPath, String cookiePath)
         {
             if (cookiePath == null)
                 return true;
             // RFC 6265, section 5.1.4, path matching algorithm.
-            if (path.equals(cookiePath))
+            if (uriPath.equals(cookiePath))
                 return true;
-            if (path.startsWith(cookiePath))
-                return cookiePath.endsWith("/") || path.charAt(cookiePath.length()) == '/';
+            if (uriPath.startsWith(cookiePath))
+                return cookiePath.endsWith("/") || uriPath.charAt(cookiePath.length()) == '/';
```

Comment:
This looks like a potential `IndexOutOfBoundsException` with the `charAt` call.

What if the parameters are ...

```
uriPath = "/path"
cookiePath = "/path-foo"
```